codec-http/src/main/java/io/netty/handler/codec/http/router/RouteResult.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2014 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http.router;
+
+import io.netty.handler.codec.http.HttpMethod;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/** Result of calling {@link Router#route(HttpMethod, String)}. */
+public class RouteResult<T> {
+    private final T                         target;
+    private final Map<String, String>       pathParams;
+    private final Map<String, List<String>> queryParams;
+
+    public RouteResult(T target, Map<String, String> pathParams, Map<String, List<String>> queryParams) {
+        this.target      = target;
+        this.pathParams  = pathParams;
+        this.queryParams = queryParams;
+    }
+
+    public T target() {
+        return target;
+    }
+
+    public Map<String, String> pathParams() {
+        return pathParams;
+    }
+
+    public Map<String, List<String>> queryParams() {
+        return queryParams;
+    }
+
+    //----------------------------------------------------------------------------
+    // Utilities to get params.
+
+    /**
+     * Extracts the first matching param in {@code queryParams}.
+     *
+     * @return {@code null} if there's no match
+     */
+    public String queryParam(String name) {
+        List<String> values = queryParams.get(name);
+        return (values == null)? null : values.get(0);
+    }
+
+    /**
+     * Extracts the param in {@code pathParams} first, then falls back to the first matching
+     * param in {@code queryParams}.
+     *
+     * @return {@code null} if there's no match
+     */
+    public String param(String name) {
+        String pathValue = pathParams.get(name);
+        return (pathValue == null)? queryParam(name) : pathValue;
+    }
+
+    /**
+     * Extracts all params in {@code pathParams} and {@code queryParams} matching the name.
+     *
+     * @return Empty list if there's no match
+     */
+    public List<String> params(String name) {
+        List<String> values = queryParams.get(name);
+        if (values == null) {
+            values = new ArrayList<String>();
```

Comment:
Check first if value != null as otherwise you can just return `Collections.emptyList()`.